lib/open_feature/hook_context.rb

```diff
@@ -0,0 +1,27 @@
+# typed: strict
+# frozen_string_literal: true
+
+module OpenFeature
+  # See https://openfeature.dev/specification/sections/hooks#41-hook-context
+  # See Requirement 4.1.1, 4.1.3, 4.1.4
+  # TODO: Requirement 4.1.2
+  class HookContext < T::Struct
+    extend T::Generic
+    extend T::Sig
+
+    include T::Struct::ActsAsComparable
+
+    Value = type_member
+    const :flag_key, String
+    const :flag_type, String
+    const :evaluation_context, EvaluationContext
+    const :default_value, Value
+
+    # Needed as opposed to .with due to https://sorbet.org/docs/tstruct#from_hash-gotchas
+    sig { params(new_context: EvaluationContext).returns(HookContext[Value]) }
+    def with_new_evaluation_context(new_context)
```

Comment:
We're trying to solve [4.1.3](https://openfeature.dev/specification/sections/hooks#requirement-414) here, correct?

I don't think we need to handle the mutation here. Rather, we need to let `before` hooks return the evaluation context where it could have optionally mutated it. We'd then have a check there to make sure no one is calling `merge` on `EvaluationContext`s in a hook that's not a `before`.

Reply:
nah it's this requirement https://openfeature.dev/specification/sections/hooks/#requirement-433

Although this here is just a helper method really. In theory, it should be possible to call `context.with(evaluation_context : new_context)` but this only works in sorbet for simple types.